src/v/cluster/rm_stm.cc

```diff
@@ -1986,6 +1986,38 @@ void rm_stm::fill_snapshot_wo_seqs(T& snapshot) {
 }
 
 ss::future<stm_snapshot> rm_stm::take_snapshot() {
+    auto start_offset = _raft->start_offset();
+
+    std::vector<abort_index> abort_indexes;
+    std::vector<abort_index> expired_abort_indexes;
+    abort_indexes.reserve(_log_state.abort_indexes.size());
+
+    for (auto& idx : _log_state.abort_indexes) {
+        if (idx.last < start_offset) {
+            // caching expired indexes instead of removing them as we go
+            // to avoid giving a control to another coroutine and managing
+            // concurrent access to _log_state.abort_indexes
+            expired_abort_indexes.push_back(idx);
+        } else {
+            abort_indexes.push_back(idx);
+        }
+    }
+    _log_state.abort_indexes = std::move(abort_indexes);
+
+    for (auto& idx : expired_abort_indexes) {
+        auto filename = fmt::format("abort.idx.{}.{}", idx.first, idx.last);
+        co_await _abort_snapshot_mgr.remove_snapshot(filename);
+    }
+
+    std::vector<tx_range> aborted;
+    aborted.reserve(_log_state.aborted.size());
+    std::copy_if(
+      _log_state.aborted.begin(),
+      _log_state.aborted.end(),
+      std::back_inserter(aborted),
+      [start_offset](tx_range range) { return range.last >= start_offset; });
+    _log_state.aborted = std::move(aborted);
+
```

Comment:
We can just fold this check into the loop in L2029, right? (ignore any range that falls under start_offset).

Reply:
L2029 is under condition, if we wait until it hit we'll keep unused indexes longer